src/metabase/models/serialization.clj

```diff
@@ -1482,6 +1509,65 @@
+    (= model :model/Field) {::fk true :export *export-field-fk* :import *import-field-fk*}
+    field-name             {::fk    true
+                            :export #(*export-fk-keyed* % model field-name)
+                            :import #(*import-fk-keyed* % model field-name)}
+    :else                  {::fk true :export #(*export-fk* % model) :import #(*import-fk* % model)}))
+
+(defn nested "Nested entities" [model backward-fk opts]
+  (let [model-name (name model)
+        sorter     (:sort-by opts :created_at)
+        key-field  (:key-field opts :entity_id)]
+    {::nested     true
+     :model       model
+     :backward-fk backward-fk
+     :export      (fn [data]
+                    ;; if this looks weird, see :series hydration for DashboardCard, it supplies Cards while we need
+                    ;; to serialize DashboardCardSeries instances
+                    (when-not (every? #(t2/instance-of? model %) data)
+                      (assert false
+                              (format "Nested data is expected to be a %s, not %s" model (t2/model (first data)))))
+                    ;; `nil? data` check is for `extract-one` case in tests; make sure to add empty vectors in
+                    ;; `extract-query` implementations for nested collections
+                    (try
+                      (->> (or data (when (nil? data)
+                                      (t2/select model backward-fk (:id *current*))))
+                           (sort-by sorter)
+                           (mapv #(extract-one model-name opts %)))
+                      (catch Exception e
+                        (throw (ex-info (format "Error exporting nested %s" model)
+                                        {:model     model
+                                         :parent-id (:id *current*)}
+                                        e)))))
+     :import      (fn [lst]
+                    (let [parent-id (:id *current*)
+                          loaded    (for [[idx ingested] (map-indexed vector lst)
+                                          ;; it's not always entity-id though, not for DashcardCardSeries
+                                          :let           [entity-id (get ingested key-field)]]
+                                      (let [ingested (assoc ingested
+                                                            backward-fk  parent-id
+                                                            :position    idx
+                                                            :serdes/meta [{:model model-name :id entity-id}])
+                                            local    (load-find-local (:serdes/meta ingested))]
+                                        (load-one! ingested local)))]
+                      (if-not (seq loaded)
+                        (t2/delete! model backward-fk parent-id)
+                        (t2/delete! model backward-fk parent-id :id [:not-in (map :id loaded)]))))}))
```

Comment:
I'm a little concerned it's possible to violate constraints enforced by toucan hooks or the database by inserting the ingested entities first, and then deleting the non-inserted entities second. I feel like deletion should come first, then inserting second. For example, there's a database constraint `unique_dimension_field_id` that enforces a 1-1 relationship between a Field and a Dimension. We'd run into that constraint here during serialization if we tried to import a Dimension for a Field, where the imported Dimension had a different `entity_id` to the existing local Dimension. I don't think this is currently covered by tests so is probably an existing bug.

Reply:
Oh, good catch!